src/mapgen/mapgen.cpp

```diff
@@ -514,15 +514,48 @@ void Mapgen::propagateSunlight(v3s16 nmin, v3s16 nmax, bool propagate_shadow)
 	//printf("propagateSunlight: %dms\n", t.stop());
 }
 
+// findPropagationCutoff
+//
+// Returns the highest y that needs light propagation.
+//
+// Above ground, there is generally a tall tower of open air with only sunlight.
+// There's no need to initiate propagation from these nodes, since they are
+// surrounded by sunlight.
+//
+// This function can skip over these open air nodes using only a single memory
+// access per node. In contrast, spreadLight() requires 7 memory accesses
+// (the node + 6 dirs) to initiate propagation.
+//
+// This speeds up calcLighting by a factor of ~2.6x (measured on X86-64).
+//
+int Mapgen::findPropagationCutoff(const v3s16 &nmin, const v3s16 &nmax)
+{
+	VoxelArea a(nmin, nmax);
+	for (int y = a.MaxEdge.Y; y >= a.MinEdge.Y; y--) {
```

Comment:
I think something like this would be useful here just to sum it up
`// Find the first node that is not sunlit`